src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -20,17 +20,104 @@ internal sealed class SocketAsyncContext
     {
         private abstract class AsyncOperation
         {
+            private enum State
+            {
+                Waiting = 0,
+                Running = 1,
+                Complete = 2,
+                Cancelled = 3
+            }
+
+            private int _state; // Actually AsyncOperation.State
+
             public AsyncOperation Next;
+            protected object CallbackOrEvent;
             public SocketError ErrorCode;
             public byte[] SocketAddress;
             public int SocketAddressLen;
 
+            public ManualResetEventSlim Event { set { CallbackOrEvent = value; } }
+
             public AsyncOperation()
             {
+                _state = (int)State.Waiting;
                 Next = this;
             }
 
-            public abstract void Complete();
+            public void QueueCompletionCallback()
+            {
+                Debug.Assert(!(CallbackOrEvent is ManualResetEventSlim));
+                Debug.Assert(_state != (int)State.Cancelled);
+
+                ThreadPool.QueueUserWorkItem(o => ((AsyncOperation)o).DoCallback(), this);
+            }
+
+            public bool TryComplete(int fileDescriptor)
+            {
+                Debug.Assert(_state == (int)State.Waiting);
+
+                return DoTryComplete(fileDescriptor);
+            }
+
+            public bool TryCompleteAsync(int fileDescriptor)
+            {
+                int state = Interlocked.CompareExchange(ref _state, (int)State.Running, (int)State.Waiting);
+                if (state == (int)State.Cancelled)
+                {
+                    // This operation has been cancelled.
+                    return true;
+                }
+                Debug.Assert(state != (int)State.Complete && state != (int)State.Running);
+
+                if (DoTryComplete(fileDescriptor))
+                {
+                    var @event = CallbackOrEvent as ManualResetEventSlim;
+                    if (@event != null)
+                    {
+                        @event.Set();
+                    }
+                    else
+                    {
+                        QueueCompletionCallback();
+                    }
+                    Volatile.Write(ref _state, (int)State.Complete);
+                    return true;
+                }
+
+                Volatile.Write(ref _state, (int)State.Waiting);
+                return false;
+            }
+
+            public bool Wait(int timeout)
+            {
+                if (((ManualResetEventSlim)CallbackOrEvent).Wait(timeout))
+                {
+                    return true;
+                }
+
+                for (;;)
+                {
+                    int state = Interlocked.CompareExchange(ref _state, (int)State.Cancelled, (int)State.Waiting);
+                    switch ((State)state)
+                    {
+                        case State.Running:
+                            // A completion attempt is in progress. Keep busy-waiting.
```

Comment:
This is a very tight spin loop.  Is this guaranteed to complete really quickly?  Consider using SpinWait.